### PR TITLE
CODENVY-457: fix NPE when swarm doesn't detect container death

### PR DESCRIPTION
**Pull request by garagatyi**
@skabashnyuk Please review

### CODE REVIEW

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java [hunk 1]
```diff
@@ -208,16 +205,13 @@ public String projectsRoot() {
...
-                    addRefAndUrlToServers(getServersWithFilledPorts(containerHost,
-                                                                    info.getNetworkSettings().getPorts()),
-                                          info.getConfig().getLabels()));
-        } catch (Exception e) {
-            // Hack needed to investigate why swarm returns invalid response
-            LOG.error(info.toString());
-            throw e;
-        }
+        Map<String, List<PortBinding>> ports = info.getNetworkSettings().getPorts();
+        Map<String, String> labels = info.getConfig().getLabels();
```
Comment (skabashnyuk) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java]: are you sure getNetworkSettings and getConfig are not null?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java [hunk 2]
```diff
@@ -208,16 +205,13 @@ public String projectsRoot() {
...
+
+        return addDefaultReferenceForServersWithoutReference(
+                addRefAndUrlToServers(getServersWithFilledPorts(containerHost,
+                                                                ports != null ? ports : Collections.emptyMap()),
```
Comment (skabashnyuk) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java]: firstNotNull?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java (hunk 1 above)
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java]: Looks like they are never null whereas ports field is null when container is not running

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java (hunk 2 above)
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java]: done

**Comment by skabashnyuk** [comment]
ok

**Comment by garagatyi** [comment]
@skabashnyuk Please take a look

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java [hunk 3]
```diff
@@ -208,16 +223,19 @@ public String projectsRoot() {
...
-        try {
-            return addDefaultReferenceForServersWithoutReference(
-                    addRefAndUrlToServers(getServersWithFilledPorts(containerHost,
-                                                                    info.getNetworkSettings().getPorts()),
-                                          info.getConfig().getLabels()));
-        } catch (Exception e) {
-            // Hack needed to investigate why swarm returns invalid response
-            LOG.error(info.toString());
-            throw e;
+        Map<String, List<PortBinding>> ports = Collections.emptyMap();
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java]: is that Collections.emptyMap shouldn't be created in the else block ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java [hunk 4]
```diff
@@ -208,16 +223,19 @@ public String projectsRoot() {
...
+        if (info.getNetworkSettings() != null && info.getNetworkSettings().getPorts() != null) {
+            ports = info.getNetworkSettings().getPorts();
+        }
+        Map<String, String> labels = Collections.emptyMap();
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java]: is that Collections.emptyMap shouldn't be created in the else block ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java (hunk 3 above)
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java]: done

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java (hunk 4 above)
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java]: done

**Comment by skabashnyuk** [comment]
ok

**Comment by benoitf** [comment]
will it be backported to 4.3.x as well ? replacing/updating https://github.com/eclipse/che/pull/1312

**Comment by garagatyi** [comment]
yes/yes

**Comment by benoitf** [comment]
nice to see this NPE gone

**Pull request merged**